core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/StepTracker.kt

```diff
@@ -0,0 +1,124 @@
+    // If copying this is too expensive, it may be changed to an AppendOnlyLinkedList.
+    // But the list should be small.
+    private val reachedSteps: MutableList<LocatedStep> = mutableListOf()
+    private val nSeenSteps: Int
+        get() = reachedSteps.size
+
+    var nStepsExcludingLookahead: Int = 0
+        private set
+
+    // Used to compute path offsets
+    private var currentPathOffset: Offset<TravelledPath> = Offset.zero()
+
+    /** Returns all the steps that have been passed on the path, in order. */
+    fun getAllReachedSteps(): List<LocatedStep> {
+        return reachedSteps
+    }
+
+    /** True if the last step has been encountered (including lookahead). */
+    fun hasSeenDestination(): Boolean {
+        return nSeenSteps == inputSteps.size
+    }
+
+    /** True if the last step has been reached, with full simulation and no lookahead. */
+    fun hasReachedDestination(): Boolean {
+        assert(nStepsExcludingLookahead <= inputSteps.size)
+        return hasSeenDestination() && nStepsExcludingLookahead == inputSteps.size
+    }
+
+    /**
+     * Go through a block range and register every step in the given range (boundaries included).
+     * New steps are first part of the "lookahead" section, and only move to the "visited" section
+     * upon `moveForward` calls.
+     */
+    fun exploreBlockRange(
+        block: BlockId,
+        rangeStart: Offset<Block>,
+        rangeEnd: Offset<Block>, // No default value as we need the infra to know the block len
+    ): List<LocatedStep> {
+        val res = mutableListOf<LocatedStep>()
+
+        // Start of the range to explore. As we pass steps, we drop the range section before the
+        // visited step (hence mutable). This ensures we don't accep out-of-order steps when they're
+        // on the same block.
+        var mutRangeStart = rangeStart
+
+        for (step in inputSteps.subList(nSeenSteps, inputSteps.size)) {
+            val location =
+                step.locations
+                    .filter { it.edge == block }
+                    .filter { it.offset in mutRangeStart..rangeEnd }
+                    .minByOrNull { it.offset } ?: break
+            val newStep =
+                LocatedStep(
+                    currentPathOffset + (location.offset - mutRangeStart),
+                    location,
+                    step,
+                )
+            res.add(newStep)
+            reachedSteps.add(newStep)
+            currentPathOffset += (location.offset - mutRangeStart)
+            mutRangeStart = location.offset
+        }
+        currentPathOffset += (rangeEnd - mutRangeStart)
+        return res
+    }
+
+    /** Integrate a part of the lookahead into the "actually visited" steps. */
+    fun moveForward(block: BlockId, start: Offset<Block>, end: Offset<Block>) {
+        nStepsExcludingLookahead +=
+            reachedSteps
+                .subList(nStepsExcludingLookahead, reachedSteps.size)
```

Comment:
Nice refactor, thanks!
nit:
```suggestion
            reachedSteps
                .drop(nStepsExcludingLookahead)
```

Reply:
Fixed here and for other `.subList` in the file